lib/Option/SanitizerOptions.cpp

```diff
@@ -34,6 +34,44 @@ static StringRef toStringRef(const SanitizerKind kind) {
   llvm_unreachable("Unsupported sanitizer");
 }
 
+SanitizerCoverageKind swift::parseSanitizerCoverageArgValue(
+    const llvm::opt::Arg *A, const llvm::Triple &Triple,
+    DiagnosticEngine &Diags, SanitizerKind sanitizer) {
+
+  SanitizerCoverageKind kind = SanitizerCoverageKind::None;
+  assert(A->getNumValues() == 1);
+
+  // The coverage names here follow the names used by clang's
+  // ``-fsanitize-coverage=`` flag.
+  kind = llvm::StringSwitch<SanitizerCoverageKind>(A->getValue(0))
+             .Case("func", SanitizerCoverageKind::Function)
+             .Case("bb", SanitizerCoverageKind::BB)
+             .Case("edge", SanitizerCoverageKind::Edge)
+             .Default(SanitizerCoverageKind::None);
+
+  if (kind == SanitizerCoverageKind::None) {
+    Diags.diagnose(SourceLoc(), diag::error_unsupported_option_argument,
+                   A->getOption().getPrefixedName(), A->getValue(0));
+    return kind;
+  }
+
+  // Running the sanitizer coverage pass will add undefined symbols to
+  // functions in compiler-rt's "sanitizer_common". "sanitizer_common" isn't
+  // shipped as a separate library we can link with. However those are defined
+  // in the various sanitizer runtime libraries so we require that we are
+  // doing a sanitized build so we pick up the required functions during
+  // linking.
+  if (kind != SanitizerCoverageKind::None && sanitizer == SanitizerKind::None) {
+    SmallString<128> optionUsed;
+    optionUsed.append(A->getOption().getPrefixedName());
+    optionUsed.append(A->getValue(0));
+    Diags.diagnose(SourceLoc(), diag::error_argument_not_allowed_with,
+                   optionUsed,
+                   "Sanitizer disabled. Please use with -sanitize=");
```

Comment:
I think we should just make a new diagnostic for this, rather than trying to cram the information into an existing diagnostic.